state_test.go

```diff
@@ -24,68 +71,307 @@ func TestState_FaultyNodes(t *testing.T) {
 		}
-		msg.View.Round = r
-		return msg
 	}
 
-	// -- test committed messages --
-	c.addMessage(msg("A", MessageReq_Commit))
-	c.addMessage(msg("B", MessageReq_Commit))
-	c.addMessage(msg("B", MessageReq_Commit))
+	maxRound, found := s.maxRound()
+	assert.Equal(t, uint64(5), maxRound)
+	assert.Equal(t, true, found)
+}
 
-	assert.Equal(t, c.numCommitted(), 2)
+func TestState_MaxRound_NotFound(t *testing.T) {
+	validatorsCount := 7
+	s := newState()
 
-	// -- test prepare messages --
-	c.addMessage(msg("C", MessageReq_Prepare))
-	c.addMessage(msg("C", MessageReq_Prepare))
-	c.addMessage(msg("D", MessageReq_Prepare))
+	validatorIds := make([]string, validatorsCount)
+	for i := 0; i < validatorsCount; i++ {
+		validatorIds[i] = fmt.Sprintf("validator_%d", i)
+	}
+	s.validators = newMockValidatorSet(validatorIds)
+
+	// Send wrong message type from some validator, whereas roundMessages map is empty
+	s.addMessage(createMessage(validatorIds[0], MessageReq_Preprepare))
+
+	maxRound, found := s.maxRound()
+	assert.Equal(t, maxRound, uint64(0))
+	assert.Equal(t, found, false)
+
+	// Seed insufficient "RoundChange" messages count, so that maxRound isn't going to be found
+	for round := range validatorIds {
+		if round%2 == 0 {
+			// Each even round should populate more than one "RoundChange" messages, but just enough that we don't reach census (max faulty nodes+1)
+			for i := 0; i < s.MaxFaultyNodes(); i++ {
+				s.addMessage(createMessage(validatorIds[mrand.Intn(validatorsCount)], MessageReq_RoundChange, uint64(round)))
+			}
+		} else {
+			s.addMessage(createMessage(validatorIds[mrand.Intn(validatorsCount)], MessageReq_RoundChange, uint64(round)))
+		}
+	}
 
-	assert.Equal(t, c.numPrepared(), 2)
+	maxRound, found = s.maxRound()
+	assert.Equal(t, uint64(0), maxRound)
+	assert.Equal(t, false, found)
 }
 
+func TestState_AddRoundMessage(t *testing.T) {
+	s := newState()
+	validatorIds := []string{"validator_1", "validator_2"}
+	s.validators = newMockValidatorSet(validatorIds)
+
+	roundMessageSize := s.AddRoundMessage(createMessage(validatorIds[0], MessageReq_Commit, 0))
+	assert.Equal(t, 0, roundMessageSize)
+	assert.Equal(t, 0, len(s.roundMessages))
+
+	s.AddRoundMessage(createMessage(validatorIds[0], MessageReq_RoundChange, 0))
+	s.AddRoundMessage(createMessage(validatorIds[0], MessageReq_RoundChange, 1))
+	s.AddRoundMessage(createMessage(validatorIds[0], MessageReq_RoundChange, 2))
+
+	roundMessageSize = s.AddRoundMessage(createMessage(validatorIds[1], MessageReq_RoundChange, 2))
+	assert.Equal(t, 2, roundMessageSize)
+
+	s.AddRoundMessage(createMessage(validatorIds[1], MessageReq_RoundChange, 3))
+	assert.Equal(t, 4, len(s.roundMessages))
+
+	assert.Empty(t, s.prepared)
+	assert.Empty(t, s.committed)
+}
+
+func TestState_addPrepared(t *testing.T) {
+	s := newState()
+	validatorIds := []string{"validator_1", "validator_2"}
+	s.validators = newMockValidatorSet(validatorIds)
+
+	s.addPrepared(createMessage(validatorIds[0], MessageReq_Commit))
```

Comment:
In most of the tests you do use ``validatorIds[0]`` to refer to one of the validators. Personally, this is a bit confusing for me because I have to double check the id of the validators all the time.

I would prefer to use simple names (A, B, C...) and then writing down the code directly.